hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ContainerKeyMapperHelper.java

```diff
@@ -390,66 +487,51 @@ private static void writeToTheDB(Map<ContainerKeyPrefix, Integer> containerKeyMa
+                                        Map<Long, AtomicLong> sharedContainerKeyCountMap,
                                         ReconContainerMetadataManager reconContainerMetadataManager)
       throws IOException {
 
-    long containerCountToIncrement = 0;
-
     for (OmKeyLocationInfoGroup omKeyLocationInfoGroup : omKeyInfo.getKeyLocationVersions()) {
       long keyVersion = omKeyLocationInfoGroup.getVersion();
       for (OmKeyLocationInfo omKeyLocationInfo : omKeyLocationInfoGroup.getLocationList()) {
         long containerId = omKeyLocationInfo.getContainerID();
         ContainerKeyPrefix containerKeyPrefix = ContainerKeyPrefix.get(containerId, key, keyVersion);
 
-        if (reconContainerMetadataManager.getCountForContainerKeyPrefix(containerKeyPrefix) == 0
-            && !containerKeyMap.containsKey(containerKeyPrefix)) {
+        // During reprocess, tables are empty so skip DB lookup - just check in-memory map
+        if (!localContainerKeyMap.containsKey(containerKeyPrefix)) {
           // Save on writes. No need to save same container-key prefix mapping again.
-          containerKeyMap.put(containerKeyPrefix, 1);
-
-          // Check if container already exists; if not, increment the count
-          if (!reconContainerMetadataManager.doesContainerExists(containerId)
-              && !containerKeyCountMap.containsKey(containerId)) {
-            containerCountToIncrement++;
-          }
-
-          // Update the count of keys for the given containerID
-          long keyCount = containerKeyCountMap.getOrDefault(containerId,
-              reconContainerMetadataManager.getKeyCountForContainer(containerId));
+          localContainerKeyMap.put(containerKeyPrefix, 1);
 
-          containerKeyCountMap.put(containerId, keyCount + 1);
+          // Thread-safe increment using computeIfAbsent (cross-task safe: FSO + OBS)
+          sharedContainerKeyCountMap.computeIfAbsent(containerId, k -> new AtomicLong(0))
+              .incrementAndGet();
         }
       }
     }
-
-    if (containerCountToIncrement > 0) {
-      reconContainerMetadataManager.incrementContainerCountBy(containerCountToIncrement);
-    }
+    // Container count will be written once at the end of reprocess, not here (Derby optimization)
   }
 
   public static boolean flushAndCommitContainerKeyInfoToDB(
-      Map<ContainerKeyPrefix, Integer> containerKeyMap,
-      Map<Long, Long> containerKeyCountMap,
+      Map<ContainerKeyPrefix, Integer> localContainerKeyMap,
+      Map<Long, AtomicLong> sharedContainerKeyCountMap,
       ReconContainerMetadataManager reconContainerMetadataManager) {
 
     try {
       // No deleted container list needed since "reprocess" only has put operations
-      writeToTheDB(containerKeyMap, containerKeyCountMap, Collections.emptyList(), reconContainerMetadataManager);
+      writeToTheDB(localContainerKeyMap, sharedContainerKeyCountMap, Collections.emptyList(), reconContainerMetadataManager);
```

Comment:
only last task getting executed need flush and cleanup map, else data can got wrong.